obs-studio-server/source/memory-manager.cpp

```diff
@@ -148,218 +155,171 @@ void MemoryManager::addCachedMemory(source_info *si)
 {
 	auto it = sources.find(obs_source_get_name(source));
-
 	if (it == sources.end())
 		return;
 
-	it->second->workers.push_back(std::thread(&MemoryManager::sourceManager, this, it->second));
+	std::unique_lock ulock(it->second->mtx);
+	it->second->workers.push_back(std::thread(&MemoryManager::sourceManager, this, std::ref(*it->second)));
 }
 
 void MemoryManager::updateSourceCache(obs_source_t *source)
 {
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
 	updateSettings(source);
 }
 
-void MemoryManager::updateSourcesCache(void)
+void MemoryManager::updateSourcesCache()
 {
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
-	for (auto data : sources)
+	for (const auto &data : sources)
 		updateSettings(data.second->source);
 }
 
-void MemoryManager::registerSource(obs_source_t *source)
+bool MemoryManager::isSourceValid(obs_source_t *source) const
 {
 	if (!source)
-		return;
+		return false;
 
 	const char *source_id = obs_source_get_id(source);
-	if (!source_id || strcmp(obs_source_get_id(source), "ffmpeg_source"))
+	if (!source_id)
+		return false;
+
+	if (strcmp(source_id, "ffmpeg_source") != 0)
+		return false;
+
+	return true;
+}
+
+void MemoryManager::registerSource(obs_source_t *source)
+{
+	if (!isSourceValid(source)) {
 		return;
+	}
 
-	std::unique_lock<std::mutex> ulock(mtx);
+	std::unique_lock ulock(mtx);
 
 	source_info *si = new source_info;
-	si->cached = false;
-	si->size = 0;
 	si->source = source;
 	sources.emplace(obs_source_get_name(source), si);
 	updateSource(source, false);
-	if (!watcher.running) {
-		watcher.running = true;
-		watcher.worker = std::thread(&MemoryManager::monitorMemory, this);
-	}
 }
 
 void MemoryManager::unregisterSource(obs_source_t *source)
 {
-	if (!source)
-		return;
-	const char *source_id = obs_source_get_id(source);
-	if (!source_id)
-		return;
-	if (strcmp(source_id, "ffmpeg_source") != 0)
+	if (!isSourceValid(source)) {
 		return;
+	}
 
-	mtx.lock();
-
-	auto it = sources.find(obs_source_get_name(source));
+	const auto *source_name = obs_source_get_name(source);
 
+	mtx.lock();
+	auto it = sources.find(source_name);
 	if (it == sources.end()) {
 		mtx.unlock();
 		return;
 	}
+
+	// Moving pointer to have a valid object when proceeding with further deinit.
+	auto moved_ptr = std::move(it->second);
+	// Removing object from the collection early to be sure that it is unavailable anymore for outer clients
+	// and someone can not spawn a new worker while we are waiting for worker threads to join.
+	// Also this prevents data race if someone called 'unregisterSource' from other thread
+	// while removal is in progress.
+	sources.erase(source_name);
 	mtx.unlock();
 
-	for (auto &worker : it->second->workers) {
+	for (auto &worker : moved_ptr->workers) {
 		if (worker.joinable())
 			worker.join();
 	}
 
-	mtx.lock();
-
-	it->second->workers.clear();
-	removeCachedMemory(it->second, true);
-
-	free(it->second);
```

Comment:
[OLD CODE]
Undefined behavior which results in memory leak.
Object was allocated with C++ `new` operator and freed with  C `free` => no destructors call for calsses like `std::vector` to free memory.